### PR TITLE
Add long issue list option 

### PR DESCRIPTION
**Pull request by vyrondrosos**

### CODE REVIEW

**Comment by dspinellis** [comment]
Thank you!  Please fix the failing test, and squash the commits into a single one.  I will then review the code.

**Comment by vyrondrosos** [comment]
I've updated my repository.

File: git-issue.sh [hunk 1]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+{
+
+  local Aid Adate Amilestone Aassignee Atags Adescription
+  #SHA id
+  Aid=$id
+  # Date
+  Adate=`git show --no-patch --format='%ai' $id`
+
+  # Milestone
+  if [ -s $path/milestone ] ; then
```
Comment (dspinellis) [inline git-issue.sh]: Please add double quotes around path

File: git-issue.sh [hunk 2]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+    Amilestone=`fmt $path/milestone`
```
Comment (dspinellis) [inline git-issue.sh]: Use the modern $( rather than `

File: git-issue.sh [hunk 3]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+  # Tags
+  if [ -s $path/tags ] ; then
+    Atags=`fmt $path/tags`
+  fi
+
+  # Description
+  Adescription=`head -n 1 $path/description` 
+
+
+  (echo -nE "$sortfield";printf '\002';echo -E "$formatstring") | #we print the field to sort by first, and remove it after sorting
```
Comment (dspinellis) [inline git-issue.sh]: Simply echo "$sortfield"$'\002'"$formatstring"

Comment (dspinellis) [inline git-issue.sh]: Write # Print the field
(Space after # and start with a capital letter.)

File: git-issue.sh [hunk 4]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+
+
+  (echo -nE "$sortfield";printf '\002';echo -E "$formatstring") | #we print the field to sort by first, and remove it after sorting 
+  sed -e s$'\003'%n$'\003'"\n"$'\003'g \
+  -e s$'\003'%i$'\003'"$Aid"$'\003'g \
+  -e s$'\003'%c$'\003'"$Adate"$'\003'g \
+  -e s$'\003'%M$'\003'"$Amilestone"$'\003'g \
+  -e s$'\003'%A$'\003'"$Aassignee"$'\003'g \
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
```
Comment (dspinellis) [inline git-issue.sh]: Place the comment before the sed invocation; Write: # An ...

File: git-issue.sh [hunk 5]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
     a)
       all=1
       ;;
+    l)
+      long=1
+      formatstring=$OPTARG
+      ;;
+    o)
+      sortfield=$OPTARG
+      if [ -z $sortfield ] ; then
```
Comment (dspinellis) [inline git-issue.sh]: Add double quotes

File: git-issue.sh [hunk 6]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
 
-  while getopts a flag ; do
+  while getopts al:o:r flag ; do
     case $flag in
     a)
       all=1
       ;;
+    l)
+      long=1
+      formatstring=$OPTARG
```
Comment (dspinellis) [inline git-issue.sh]: Add double quotes

File: git-issue.sh [hunk 7]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+      ;;
+    o)
+      sortfield=$OPTARG
+      if [ -z $sortfield ] ; then
+        usage_list
+      fi
+      if echo $sortfield | grep -v -E '^%[icMATD]$' > /dev/null; then
```
Comment (dspinellis) [inline git-issue.sh]: I think you can write this with expr(1)

File: git-issue.sh [hunk 8]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
     ?)
       usage_list
       ;;
     esac
   done
-  shift $(($OPTIND - 1));
 
+  case $formatstring in
+    oneline)
+      formatstring='ID:%i  DATE:%c  TAGS:%T  DESC:%D'
```
Comment (dspinellis) [inline git-issue.sh]: Shouldn't the words be capital case, e.g. Date?
Also space after :

File: git-issue.sh [hunk 9]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+  (echo -nE "$sortfield";printf '\002';echo -E "$formatstring") | #we print the field to sort by first, and remove it after sorting 
+  sed -e s$'\003'%n$'\003'"\n"$'\003'g \
+  -e s$'\003'%i$'\003'"$Aid"$'\003'g \
+  -e s$'\003'%c$'\003'"$Adate"$'\003'g \
+  -e s$'\003'%M$'\003'"$Amilestone"$'\003'g \
+  -e s$'\003'%A$'\003'"$Aassignee"$'\003'g \
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
+  tr "\n" '\001' >> results
+  printf "\n" >> results
```
Comment (dspinellis) [inline git-issue.sh]: echo is enough

File: git-issue.sh [hunk 10]
```diff
@@ -855,26 +855,97 @@ usage_list()
...
+
+  (echo -nE "$sortfield";printf '\002';echo -E "$formatstring") | #we print the field to sort by first, and remove it after sorting 
+  sed -e s$'\003'%n$'\003'"\n"$'\003'g \
+  -e s$'\003'%i$'\003'"$Aid"$'\003'g \
+  -e s$'\003'%c$'\003'"$Adate"$'\003'g \
+  -e s$'\003'%M$'\003'"$Amilestone"$'\003'g \
+  -e s$'\003'%A$'\003'"$Aassignee"$'\003'g \
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
+  tr "\n" '\001' >> results
```
Comment (dspinellis) [inline git-issue.sh]: Don't append do the file. Instead write to stdout, and have a single tee capture the output. (See below.)

File: git-issue.sh [hunk 11]
```diff
@@ -889,13 +960,22 @@ sub_list()
...
-  done |
-  sort -k 2 |
-  tee results |
-  pager
+  if [ "$long" ] ; then
+   
+    while read path id ; do
+      shortshow
+    done 
+    cp results ~/res
```
Comment (dspinellis) [inline git-issue.sh]: Forgotten debug statement?

File: git-issue.sh [hunk 12]
```diff
@@ -889,13 +960,22 @@ sub_list()
...
-    head -1 $path/description
-  done |
-  sort -k 2 |
-  tee results |
-  pager
+  if [ "$long" ] ; then
+   
+    while read path id ; do
+      shortshow
+    done
```
Comment (dspinellis) [inline git-issue.sh]: done |
sort $sortrev | 
sed 's/^.*\x02//' | 
tr '\001' "\n"

File: git-issue.sh [hunk 13]
```diff
@@ -889,13 +960,22 @@ sub_list()
...
+    done 
+    cp results ~/res
+    cat results | sort $sortrev | sed 's/^.*\x02//' | tr '\001' "\n" | pager 
+  else
+    while read path id ; do
+      printf '%s' "$id "
+      head -1 $path/description
+    done |
+    sort -k 2 |
+    tee results |
```
Comment (dspinellis) [inline git-issue.sh]: Move tee results | pager outside the if statement.

File: git-issue.sh [hunk 14]
```diff
@@ -985,6 +1065,7 @@ Work with an issue
 
 Show multiple issues
    list       List open issues (or all with -a)
+* list -l formatstring: This will list issues in a specifiable format, given as an argument to -l
```
Comment (dspinellis) [inline git-issue.sh]: in the specified format.

**Comment by vyrondrosos** [comment]
Thanks for the feedback. I made the changes, wrote a few tests, and squashed the commits again. However, it seems like the tests I added fail on the Travis CI system, while they pass on mine. Is that normal?

**Comment by dspinellis** [comment]
It's not normal. Only one of the two test targets fails.  Could it be that you're using a non-standard sed feature?

```
sed: -e expression #1, char 17: unknown option to `s'
fail 85 - grep Second issue
sed: -e expression #1, char 17: unknown option to `s'
ok 86 - not grep First-issue
sed: -e expression #1, char 17: unknown option to `s'
sed: -e expression #1, char 17: unknown option to `s'
fail 87 - grep feature
```

**Comment by vyrondrosos** [comment]
After doing some testing, it seems that the problem lies with the nonprintable characters I use as delimiters to `sed`(since the fields themselves, especially the issue's description, could contain any printable character). Not sure what the alternative would be though, apart from switching `sed` out for another tool.

**Comment by vyrondrosos** [comment]
The implemented tests are now passing. If you have any ideas on further improvements, please let me know.

**Comment by dspinellis** [comment]
Great!  The code review seems to be pending.

**Comment by vyrondrosos** [comment]
Indeed, it seems I had missed some of your suggestions(the ones Github didn't show automatically). It should be fixed now.

File: README.md [hunk 15]
```diff
@@ -93,6 +93,27 @@ You use _git issue_ with the following sub-commands.
 ### Show multiple issues
 * `git issue list`: List open issues (or all with `-a`).
    An optional argument can show issues matching a tags or milestone.
+* `git issue list -l formatstring`: This will list issues in the specified format, given as an argument to `-l`.
+   The following escape sequences can be used:
+   
+   - `%n` : newline
+
```
Comment (dspinellis) [inline README.md]: Can we please loose the empty lines between the bullets?

File: git-issue.sh [hunk 16]
```diff
@@ -855,26 +855,98 @@ usage_list()
...
+  date=$(git show --no-patch --format='%ai' $id)
+
+  # Milestone
+  if [ -s "$path/milestone" ] ; then
+    milestone=$(fmt $path/milestone|sed -e 's/[\/&]/\\&/g') # Escape sed special chars before passing them 
+  fi
+
+  # Assignee
+  if [ -r "$path/assignee" ] ; then
+    assignee=$(fmt $path/assignee|sed -e 's/[\/&]/\\&/g')
```
Comment (dspinellis) [inline git-issue.sh]: Some more quotes in $path are needed.

File: git-issue.sh [hunk 17]
```diff
@@ -855,26 +855,98 @@ usage_list()
...
+  # Tags
+  if [ -s "$path/tags" ] ; then
+    tags=$(fmt $path/tags|sed -e 's/[\/&]/\\&/g')
+  fi
+
+  # Description
+  description=$(head -n 1 $path/description|sed -e 's/[\/&]/\\&/g')
+
+
+  (echo "$sortfield"$'\002'"$formatstring") | # Print the field to sort by first, and remove it after sorting
```
Comment (dspinellis) [inline git-issue.sh]: Move the comment on the line above

File: git-issue.sh [hunk 18]
```diff
@@ -855,26 +855,98 @@ usage_list()
...
+    tags=$(fmt $path/tags|sed -e 's/[\/&]/\\&/g')
+  fi
+
+  # Description
+  description=$(head -n 1 $path/description|sed -e 's/[\/&]/\\&/g')
+
+
+  (echo "$sortfield"$'\002'"$formatstring") | # Print the field to sort by first, and remove it after sorting 
+
+  # Nonprintable character is used as delimiter since the fields can contain any printable character
```
Comment (dspinellis) [inline git-issue.sh]: This comment is no longer valid

**Comment by dspinellis** [comment]
Thank you; almost there!  Just a few small changes remaining.

**Comment by vyrondrosos** [comment]
Done.

**Comment by dspinellis** [comment]
Well done, thank you!

**Pull request merged**